### ISSUE TITLE
Implement label encoding

### ISSUE THREAD

**Issue by luizgh** [Menelau/DESlib#39]
Currently, the y values for the dataset are expected to be numeric, starting from 0 and with no gaps. Using it with other labels (e.g. -1 and +1) is not working. This can be addressed using sklearn's label encoder:

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html

**Comment by luizgh** [comment]
I have started implementing it (cfc32c00be7eed44c433db3c7f0854141cafd02f). The core idea is to use the sklearn's label encoder: saving the indices (enc.transform(y)) and the original class names (enc.classes_). Right now, it is working with classes from the "DSEL" dataset that are encoded (e.g. classes described as strings).

However, since we don't control de creation of the base classifiers, each base classifier will have their own label encoder, (and therefore the clf.predict() will return the original class names). So this needs to be taken into consideration. Here is the list of what I think we need to do:

- Add a check in the DS constructor: all base classifier should predict the same classes
- Check all locations where "predict" from the base classifiers is called. We need to save the label encoder from the DS method, and use it to "transform" the classifier predictions. Only at the end (when returning to the user), we should encode it back to the original classes.

@Menelau, as we discussed briefly today, we can already consider an optimization of the code, and save the predictions of all base classifiers in the "predict" method - since some methods will need these results. However, I am thinking that instead of saving these results as a class variable, maybe it is better to pass them as arguments (introducing new class variables in the "predict" method can be counter-intuitive, and introduce bugs in the future. Passing the predictions as arguments makes it an explicit dependency) - let me know what you think

**Comment by Menelau** [comment]
@luizgh I agree with you. It is not intuitive introducing new class variables on the predict method and since the predict method may be called several times during the program execution (for classifying different blocks of data) it can introduce errors.

 I believe the best option is to calculate this matrix, and pass it as an argument to the classify_instance method.

**Comment by luizgh** [comment]
Most of the functionality has been implemented in 9fec8ac739fde2d71fb85c87188edeb1562be02e. 
Pending:

From the DS base "predict" method: 
Pass predictions to classify_instance 
Pass predictions to estimate_competence

In particular, the DCS and DES algorithms should not be calling "predict" on the base classifiers, but rather use the predictions passed from the DS "predict" method

**Comment by Menelau** [comment]
I started working on passing the pre-computed predictions to the classify_instance and estimate_competence methods.

The problem with the estimate_competence is that only a few DS techniques require the prediction information for the competence estimation (LCA, MLA, A Posteriori, MCB, KNOP and META-DES). So we cannot force the same method signature for all DS classes as in the previous version

**Comment by Menelau** [comment]
I've changed the code to pass the classifier's predictions for the classify_instance and estimate_competence methods and update the test routines (commit 6efa75afeba02f9b4bd55c036083c2ebf18cf286). I believe now the label encoder now is working fine. 

For now, the estimate_competence of all DS classes are receiving the base prediction even though they do not need it. It may not be the best solution, but addressing that can be done in another issue, especially since the structure of the code may change considerably in order to process the classification in batches (issue #42)

**Comment by luizgh** [comment]
Everything seems to be working now. The last pending item is to document this on a "developer's guide". In particular, it should be mentioned that sklearn's ensembles work differently than a list of classifiers (the base classifiers in sklearn return encoded labels in their "predict" functions).
